File: push/src/main/java/com/xiaomi/xmsf/CrashHandler.java
package com.xiaomi.xmsf;

import android.widget.Toast;

import com.elvishew.xlog.Logger;
import com.elvishew.xlog.XLog;
import com.elvishew.xlog.flattener.ClassicFlattener;
import com.elvishew.xlog.printer.file.FilePrinter;
import com.elvishew.xlog.printer.file.backup.NeverBackupStrategy;
import com.elvishew.xlog.printer.file.clean.FileLastModifiedCleanStrategy;
import com.elvishew.xlog.printer.file.naming.DateFileNameGenerator;
import com.xiaomi.xmsf.utils.LogUtils;

import top.trumeet.common.utils.Utils;

public class CrashHandler {
    private static Thread.UncaughtExceptionHandler defaultHandler;

    static {
        initDefaultHandler();
    }

    public static void installCrashLogger() {
        final String TAG = CrashHandler.class.getSimpleName();
        final Logger logger = XLog.tag(TAG).build();
        final Logger crashLogger = XLog.tag(TAG).printers(createCrashPrinter()).build();

        install((t, e) -> {
            StringBuilder crashInfo = new StringBuilder();
            crashInfo.append("Mi Push Crash:\n");
            crashInfo.append(e);
            StackTraceElement[] stackTrace = e.getStackTrace();
            for (int i = 0; i < 3 && i < stackTrace.length; ++i) {
                crashInfo.append("\n    ");
                crashInfo.append(stackTrace[i]);
            }
            Utils.makeText(crashInfo, Toast.LENGTH_LONG);
            logger.e("Mi Push Crash", e);
            crashLogger.e("Mi Push Crash", e);
        });
    }

    private static FilePrinter createCrashPrinter() {
        int _7DaysInMillis = 7 * 24 * 60 * 60 * 1000;
        return new FilePrinter.Builder(LogUtils.getLogFolder(Utils.getApplication()))
                .fileNameGenerator(new DateFileNameGenerator() {
                    @Override
                    public String generateFileName(int logLevel, long timestamp) {
                        return "Crash_" + super.generateFileName(logLevel, timestamp);
                    }
                })
                .backupStrategy(new NeverBackupStrategy())
                .cleanStrategy(new FileLastModifiedCleanStrategy(_7DaysInMillis))
                .flattener(new ClassicFlattener())
                .build();
    }

    public static void install(Thread.UncaughtExceptionHandler handler) {
        Thread.setDefaultUncaughtExceptionHandler((t, e) -> {
            handler.uncaughtException(t, e);
            if (defaultHandler != null) {
                defaultHandler.uncaughtException(t, e);
            }
        });
    }

    public static void uninstall() {
        Thread.setDefaultUncaughtExceptionHandler(defaultHandler);
    }

    public static void initDefaultHandler() {
        defaultHandler = Thread.getDefaultUncaughtExceptionHandler();
    }
}

File: push/src/main/java/com/xiaomi/xmsf/CrashHandler.kt
package com.xiaomi.xmsf

import android.widget.Toast
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import io.github.aakira.napier.Napier
import com.xiaomi.xmsf.utils.LogUtils
import top.trumeet.common.utils.Utils

object CrashHandler {
    private var defaultHandler: Thread.UncaughtExceptionHandler? = null

    init {
        initDefaultHandler()
    }

    @JvmStatic
    fun installCrashLogger() {
        val tag = "CrashHandler"
        val logDir = File(LogUtils.getCrashFolder(Utils.getApplication()!!))

        install { _, e ->
            val crashInfo = StringBuilder()
            crashInfo.append("Mi Push Crash:\n")
            crashInfo.append(e)
            val stackTrace = e.stackTrace
            for (i in 0 until minOf(3, stackTrace.size)) {
                crashInfo.append("\n    ")
                crashInfo.append(stackTrace[i])
            }
            Utils.makeText(crashInfo, Toast.LENGTH_LONG)
            Napier.e("Mi Push Crash", e, tag = tag)
            writeCrashToFile(logDir, e)
        }
    }

    private fun writeCrashToFile(logDir: File, throwable: Throwable) {
        try {
            if (!logDir.exists()) logDir.mkdirs()
            val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
            val timeFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
            val fileName = "Crash_${dateFormat.format(Date())}.txt"
            val file = File(logDir, fileName)
            val time = timeFormat.format(Date())
            val line = "$time [ERROR] CrashHandler: Mi Push Crash ${throwable.stackTraceToString()}\n"
            file.appendText(line)
        } catch (_: Exception) {}
    }



    @JvmStatic
    fun install(handler: Thread.UncaughtExceptionHandler) {
        Thread.setDefaultUncaughtExceptionHandler { t, e ->
            handler.uncaughtException(t, e)
            defaultHandler?.uncaughtException(t, e)
        }
    }

    @JvmStatic
    fun uninstall() {
        Thread.setDefaultUncaughtExceptionHandler(defaultHandler)
    }

    @JvmStatic
    fun initDefaultHandler() {
        defaultHandler = Thread.getDefaultUncaughtExceptionHandler()
    }
}
